control/state.py

```diff
@@ -9,95 +9,164 @@
+    def _remove_key(self, key: str):
+        """Removes key from state data store."""
         pass
 
-    @abstractmethod
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the state data store."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._add_key(key, val)
+
+    def remove_bdev(self, bdev_name: str):
+        """Removes a bdev from the state data store."""
+        key = self.BDEV_PREFIX + bdev_name
+        self._remove_key(key)
+
+    def add_namespace(self, subsystem_nqn: str, nsid: str, val: str):
+        """Adds a namespace to the state data store."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + nsid
+        self._add_key(key, val)
+
+    def remove_namespace(self, subsystem_nqn: str, nsid: str):
+        """Removes a namespace from the state data store."""
+        key = self.NAMESPACE_PREFIX + subsystem_nqn + "_" + nsid
+        self._remove_key(key)
+
     def add_subsystem(self, subsystem_nqn: str, val: str):
-        pass
+        """Adds a subsystem to the state data store."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._add_key(key, val)
 
-    @abstractmethod
     def remove_subsystem(self, subsystem_nqn: str):
-        pass
+        """Removes a subsystem from the state data store."""
+        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
+        self._remove_key(key)
 
-    @abstractmethod
     def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
-        pass
+        """Adds a host to the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._add_key(key, val)
 
-    @abstractmethod
     def remove_host(self, subsystem_nqn: str, host_nqn: str):
-        pass
+        """Removes a host from the state data store."""
+        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
+        self._remove_key(key)
 
-    @abstractmethod
-    def add_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str,
-                     val: str):
-        pass
+    def add_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                     traddr: str, trsvcid: str, val: str):
+        """Adds a listener to the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._add_key(key, val)
 
-    @abstractmethod
-    def remove_listener(self, subsystem_nqn: str, traddr: str, trsvcid: str):
-        pass
+    def remove_listener(self, subsystem_nqn: str, gateway: str, trtype: str,
+                        traddr: str, trsvcid: str):
+        """Removes a listener from the state data store."""
+        key = "{}{}_{}_{}_{}_{}".format(self.LISTENER_PREFIX, gateway,
+                                        subsystem_nqn, trtype, traddr, trsvcid)
+        self._remove_key(key)
 
     @abstractmethod
     def delete_state(self):
```

Comment:
Is this used anywhere? Do we need it? If yes, we should make sure that there is a mechanism that someone doesn't accidentally remove the state.

Reply:
I use it for testing purposes. We can move it to a test context in a different PR.